File: code/C++/47-Permutations II.cpp

/*Permutations II 

Given a collection of numbers that might contain duplicates, return all possible unique permutations.

For example,
[1,1,2] have the following unique permutations:
[1,1,2], [1,2,1], and [2,1,1].
*/

class Solution {
public:
    vector<vector<int> > permuteUnique(vector<int> &num) {
        vector<vector<int> > result;
        unordered_map<int, int> exist;
        
        if(num.size() == 0)
          return result;
        
        if(num.size() == 1){
            
            result.push_back(num);
            return result;
        }
        
        
        //DFS
        
        int length = num.size();
        
        for(int i=length-1;i>=0;i--){
            vector<vector<int> > nextPer;
            vector<int> temp(num);
            int curNum = num[i];
            if(exist.find(curNum) != exist.end()) {
                continue;
            }
            
            else {
                exist[curNum] = 1;
            }
            
            temp.erase(temp.begin() + i);
            nextPer = permuteUnique(temp);//这条好屌, 从temp中获取所有不重复的permution
            for(int j = 0;j<nextPer.size();j++){
                nextPer[j].push_back(curNum);
                vector<int> temp_result = nextPer[j];
                result.push_back(temp_result);
            }
         }
         
         return result;
    }
};

File: code/Python/47-Permutations II.py
# -*- coding: utf-8 -*
"""
@author: Kaka Chen
@license: (C) Copyright Personal Uncommercial Limited.
@contact: kakachen1990@gmail.com
@file: 47-Permutations II.py
@time: 2020/9/18 8:22 下午
@desc:
Given a collection of numbers that might contain duplicates, return all possible unique permutations.

Example:

Input: [1,1,2]
Output:
[
  [1,1,2],
  [1,2,1],
  [2,1,1]
]
"""


class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []

        def backtrack(left, history):
            nonlocal res
            if not left:
                res.append(history)
            for i in set(left):
                left.remove(i)
                left.append(i)
                backtrack(left[:-1], history + [i])

        backtrack(nums, [])
        return res


if __name__ == '__main__':
    s = Solution()
    ipt = [1, 1, 2]
    print(s.permuteUnique(ipt))
